### PR TITLE
Add docker services for SLDB

### PR DESCRIPTION
**Pull request by badosu**

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    && mkdir -p /sldb
+
+## from tianon/perl
+ENV PERL_CPANM_OPT --verbose --mirror https://cpan.metacpan.org --mirror-only
+RUN cpanm --notest Digest::SHA Module::Signature && rm -rf ~/.cpanm
+ENV PERL_CPANM_OPT $PERL_CPANM_OPT --verify
+
+WORKDIR /sldb
+
+RUN cpanm --quiet --notest Inline::Python \
```
Comment (badosu) [inline Dockerfile:51]: It's not quiet, it's a trap!

**Comment by Yaribz** [comment]
Thank you!

A few comments/questions:
1) Am I right thinking this deployment method can only be used to migrate an existing SLDB instance (as I see no optional code to initialize the database of the sldb-db component) ?
And in this case, I guess the "Migrate the database" step detailed in the README file isn't really optional ? Or maybe your are only considering the use case of redeploying SLDB components (other than the DB) for an existing SLDB instance ?

2) I'm worried about the fact that this duplicates configuration files definitions (ratingEngine.conf, slMonitor.conf...), which means more places to update in the repository if I decide to add new settings in some conf files for example.
I'm considering updating the setup script to add support for an unattended mode which would make it possible to just generate the base conf files, or init the db programmatically for example. Also, I would provide a Perl script instead of replace-vars.sh, which would allow updating/adding parameters in conf files without requiring placeholders. That way it could work directly from the conf files generated by the setup script. This would make it possible to fix both 1. and 2.
What is your opinion about this ?

3) What is the goal of the touch command at the end of the sldb-entrypoint.sh script ? (https://github.com/badosu/SLDB/blob/badosu/dockerize/docker/scripts/sldb-entrypoint.sh#L22)

**Comment by badosu** [comment]
1. > Am I right thinking this deployment method can only be used to migrate an existing SLDB instance (as I see no optional code to initialize the database of the sldb-db component) ?

As soon as one runs `docker-compose up sldb-db` the database is initialized with the `MYSQL_` variables configuration. So yes, it's not only optional in the sense that one can use a newly provided database with fresh data but also in the sense one can avoid using the docker provisioning for the database.

2. > I'm worried about the fact that this duplicates configuration files definitions ... What is your opinion about this ?

Sure, if a different configuration method is provided I don't mind using it. The PR will still contain changes that target the currently available configuration method.

3. > What is the goal of the touch command at the end of the sldb-entrypoint.sh script ?

So that the `echo` call just below does not fail when the logfile does not exist.

**Comment by Yaribz** [comment]
Good ! I will try to perform the changes I described in SLDB to allow clean Docker integration.

**Comment by badosu** [comment]
> (regarding 3., I think you don't need it: >> should create the logfile if needed)

`>` does, `>>` does not

**Comment by Yaribz** [comment]
> `>` does, `>>` does not

Really ? On what system do you have this behavior ?

(btw I think only one `%s` is needed in the `echo` call, here: `printf '%s %s\n' "$(date)"`)

**Comment by badosu** [comment]
Well I can cleanup later, as soon as I notice there's some intent to merge as is, or if I rework this later after the configuration changes.

**Comment by Yaribz** [comment]
Sure, no problem. My plan is indeed to integrate this PR with the changes I explained above, but for this I need to update some SLDB code first, and I can't give precise delay regarding this (it all depends on my free time and my activity on other projects...).

Also, I was wondering if it was possible to put the docker files `.env`, `.dockerignore`, `Dockerfile` and `docker-compose.yml` in the docker directory, to clearly separate main SLDB files from the docker specific ones ?

**Comment by badosu** [comment]
> Also, I was wondering if it was possible to put the docker files .env, .dockerignore, Dockerfile and docker-compose.yml in the docker directory, to clearly separate main SLDB files from the docker specific ones ?

It might be possible, I'll revisit this after the PR got the green flag to proceed

**Comment by Yaribz** [comment]
After deeper review, I think the conf problem I spoke about only exists for the `sldbLi.conf` file.
Indeed, the other SLDB conf files don't need to have all parameters declared in the file, so if I add a new parameter the docker system will still work.

The `sldbLi.conf` file isn't generated by the setup script, it is directly available in the repository. So actually I think I will just implement a Perl script which is able to update/create a configuration file based on environment variables, and which doesn't require inserting placeholders in the conf file beforehand.

**Comment by Yaribz** [comment]
Ok I have added the missing `updateConfFile.pl` script.

What remains to be done (I think):
- remove the `sldbLi.conf` file contained in the pull request (the original conf file will be used instead)
- create a file (`sldbLi.conf.docker` ?) containing only the modified settings of the original `sldbLi.conf` file
- call the new `updateConfFile.pl` script from the entrypoint, to automatically update the original `sldbLi.conf` file
      (something like this I guess, with the correct paths: `perl updateConfFile.pl sldbLi.conf -f sldbLi.conf.docker`)
- remove the `sldb.conf` file which contains BAR-specific customizations (handled in BAR repository ?)
- gather all docker-specific files in the docker directory

**Comment by badosu** [comment]
I need configuration for all components, will I be able to configure all components instead of only sldbLi?

> remove the sldb.conf file which contains BAR-specific customizations (handled in BAR repository ?)

So you can't configure default privacy on vanilla spads?

**Comment by badosu** [comment]
Ok, I refactored it to use the provided configuration method.

I gotta say it's not much better than previously.

At any rate, it's usable and cleaner for the SLDB repository I guess. The only issue is that users.conf, levels.conf and commands.conf (I suppose) still require a configuration method. But I can address that in some other way

**Comment by badosu** [comment]
I marked this PR as WIP until I'm comfortable with the result and tested it.

**Comment by Yaribz** [comment]
> Perhaps makes sense for this file to exist in SimpleConf repo?

In a way, yes.

If the script was a bit more complex than just using the update conf functionality (which is entirely implemented in the `SimpleConf` module), and if I could see it used somewhere else than just SLDB, I guess I would have done it this way. For now I chose to keep `SimpeConf` as it was planned initially: just a Perl module. But you're totally right, maybe it's not the cleanest way...

> So you can't configure default privacy on vanilla spads?

Sure you can, but if I wanted to change the default privacy mode in SLDB I would just change the default value, I wouldn't provide a configuration file by default which would change the default value of the code.

**Comment by badosu** [comment]
> Sure you can, but if I wanted to change the default privacy mode in SLDB I would just change the default value

How can that be done?

**Comment by Yaribz** [comment]
> Ok, I refactored it to use the provided configuration method.
> 
> I gotta say it's not much better than previously.

Sorry I think there has been a misunderstanding. What I meant by:

> After deeper review, I think the conf problem I spoke about only exists for the sldbLi.conf file.
> Indeed, the other SLDB conf files don't need to have all parameters declared in the file, so if I add a new parameter the docker system will still work.

Is that actually I totally agree with you concerning most of the SLDB conf files. You can totally keep your previous way of handling them because it's already totally forward compatible with any optional parameters I would add in these conf files. The only file which needed to be handled through `updateConfFile.pl` was `sldbLi.conf`, because this file doesn't support optional parameters. So if I add a new parameter in the future your current system would break until someone add the same parameter in the `sldbLi.conf` file you provide.

**Comment by Yaribz** [comment]
> > Sure you can, but if I wanted to change the default privacy mode in SLDB I would just change the default value
> 
> How can that be done?

By editing the code...

What I mean is: your file is overriding SLDB's default (changing default `privacyMode` from `1` to `0`). This is fine for BAR, because this is the configuration BAR wants. But this is not fine for SLDB, because SLDB already provides default values. If I wanted SLDB to have different default values I would just update the code.

**Comment by badosu** [comment]
> What I mean is: your file is overriding SLDB's default (changing default privacyMode from 1 to 0). This is fine for BAR, because this is the configuration BAR wants. But this is not fine for global SLDB repository, because SLDB already provides default values  If I wanted SLDB to have different default values I would just update the code.

Ok, I think there was some miscommunication then. The value is configurable, you just don't want it exposed. By the way, I think having the configuration files with the defaults exposed present on the repository is good not only for completeness but for newcomers understand their options. In any case, I don't care too much about it and I assume you don't either since ideally the managed hosting method will use the setupsldb script.

> The only file which needed to be handled through updateConfFile.pl was sldbLi.conf, because this file is actually already  included in SLDB repository (not generated by the setup script) and it doesn't support optional parameters.

It's ok, I think this works for me, I prefer to have all values consistent anyway

**Comment by badosu** [comment]
Ok, I sent a last commit to changes I think are close to complete configurability.

Next step is battle testing it, in which success I'll update the README and mark as ready for review

**Comment by Yaribz** [comment]
> The value is configurable, you just don't want it exposed.

On the contrary, if I had more time I would love having all SLDB configuration settings fully documented...

Sorry, it seems I really didn't manage to explain clearly what I meant, I will try again:
SLDB is supposed to have a default configuration which should be consistent whether you are installing it using the setup script or through Docker. Imo there is no reason one should have a different default `privacyMode` value when using Docker than when using the setup script. Currently, with your PR, someone installing using Docker will have a `sldb.conf` file and will have a default `privacyMode` set to `0`, whereas someone installing using the setup script will not have a `sldb.conf` file and will have a default `privacyMode` set to `1`. That's the only thing which bothers me.

> By the way, I think having the configuration files with the defaults exposed present on the repository is good not only for completeness but for newcomers understand their options. In any case, I don't care too much about it and I assume you don't either since ideally the managed hosting method will use the setupsldb script.

I understand. Current situation is clearly not ideal anyway because:
1) all settings aren't documented
2) all settings aren't listed in the conf files by default as you said

If at least one of these two statements were false it would be better...

**Comment by badosu** [comment]
> Currently, with your PR, someone installing using Docker will have a sldb.conf file and will have a default privacyMode set to 0, whereas someone installing using the setup script will not have a sldb.conf file and will have a default privacyMode set to 1.

I understand it, that's why I set the it as 1 in my sample configuration. I think it's just a matter of words, if you just had said to set it to 1 to reflect the true default value on sldb, instead of removing the config altogether, it would have been clearer for me :-)

In any matter, next step is testing it in production :cowboy_hat_face:

**Comment by Yaribz** [comment]
> I understand it, that's why I change it to 1 in my sample configuration. I think it's just a matter of words, if you just had said to set it to 1 to reflect the true default value on sldb, instead of removing the config altogether, it would have been clearer for me :-)

Tbh I think it's better to also have consistency in the existence of the `sldb.conf` file (still for the same reason: if I decide to change the default value in the future for example...). But I guess it's less problematic as long as I don't change defautl values...

> In any matter, next step is testing it in production 🤠

Thanks for your work !

Btw, just to be sure I explained correctly the `updateConfFile.pl` script: it is able to read values from files and to interpret environment variables placeholders. This means for example you can put a `sldbLi.conf.docker` file in your docker container with this content:
```sldbLi%lobbyLogin:${LOBBY_LOGIN}
lobbyPassword:${LOBBY_PASSWORD}
lobbyHost:${LOBBY_HOST}
lobbyPort:${LOBBY_PORT}
sldb:${SLDB_DB_USER}/${SLDB_DB_PASSWORD}@${SLDB_DB_DATABASE}
```
And then call the `updateConfFile.pl` like this:
`perl updateConfFile.pl etc/sldbLi.conf -f sldbLi.conf.docker`

The `updateConfFile.pl` script will first read the content of the `sldbLi.conf.docker` file. Then, **_in memory_**, it will replace the placeholders by the current values of the corresponding environment variables, and then it will apply the changes to the `sldbLi.conf` file.

**Comment by badosu** [comment]
Yes, I got the `--from-file` usage, that was my first intention to use. However I'd prefer to just grab whatever is in `etc` at the root SLDB directory and reconfigure it, and if not present to create config from scratch. It allows for flexibility as in the case SLDB repo starts including default distribution etc files for the other components as well.

**Comment by Yaribz** [comment]
> Yes, I got the `--from-file` usage, that was my first intention to use. However I'd prefer to just grab whatever is in `etc` at the root SLDB directory and reconfigure it, and if not present to create config from scratch. It allows for flexibility as in the case SLDB repo starts including default distribution etc files for the other components as well.

Yes, but all this is totally compatible with the `--from-file` usage.

> I had the provisional partial `.conf` file, but it didn't look too good for me.

Oh Ok. I though You'd rather having clean separate conf files as initially, compared to having exactly the same contents gathered in a single line for each component in your main compose file. All is ok then if it's a deliberate choice :)

**Comment by badosu** [comment]
@Yaribz Updated: 

- Included default files for simplicity and ease of operation, flexibility is provided via `<COMPONENT>_CONF` variables
- Added initSldb script
- Extended README.md

**Comment by Yaribz** [comment]
>  Included default files for simplicity and ease of operation, flexibility is provided via `<COMPONENT>_CONF` variables

Nice !

>  Added initSldb script

Hum, I would like to avoid code duplication as much as possible. I could try to replace this script by an automatic mode in the setup script that you could use, something like `./setupSldb.pl --init-db` ?

**Comment by badosu** [comment]
> Hum, I would like to avoid code duplication as much as possible. I could try to replace this script by an automatic mode in the setup script that you could use, something like ./setupSldb.pl --init-db ?

This is sensible, as of now I'm using initSldb as a docker helper script for it's specific considerations.

I can refactor later as something general supported by the main initSldb script.

**Comment by badosu** [comment]
> I added the `--init-db` option to the setup script (commit [2369e74](https://github.com/Yaribz/SLDB/commit/2369e74ef6924f9b5014308a44c812ef5addb868))

One reason I preferred to make a local script is that I need the sqlfile present either in the log or etc directory since those are the only shared volumes I present to the user via docker.

One alternative is for an option to print the statements instead of sqlfile.

**Comment by Yaribz** [comment]
> One alternative is for an option to print the statements instead of sqlfile.

The latest version of the setup script (including the fix for the `DBI:` syntax) prints the SQL statements when it detects the database isn't local, maybe it handles your use case ?

**Comment by badosu** [comment]
Awesome, thanks!

**Comment by badosu** [comment]
@Yaribz This is ready for review and running in production now, I'll wait a day to confirm it's safe to merge

**Comment by Yaribz** [comment]
It's ok for me too, except I don't understand why [this commit](https://github.com/badosu/SLDB/commit/174062059f4ad3f3e1274ea6ecf49a284f2f4660#diff-48724c713183d9a74fa3e7793a01485e985137ec1da6a7470f2c068cbb5f8d31) of the PR shows modification of an empty `updateConfFile.pl` file ?
I think it will make the file history a bit weird after the PR merge...

**Comment by badosu** [comment]
> It's ok for me too, except I don't understand why [this commit](https://github.com/badosu/SLDB/commit/174062059f4ad3f3e1274ea6ecf49a284f2f4660#diff-48724c713183d9a74fa3e7793a01485e985137ec1da6a7470f2c068cbb5f8d31) of the PR shows modification of an empty `updateConfFile.pl` file ? I think it will make the file history a bit weird after the PR merge...

Makes it executable, so it's consistent with the other components.

**Pull request merged**